blssig/src/bdn/mod.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019-2024 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//! BDN (Boneh-Drijvers-Neven) signature aggregation scheme, for preventing rogue public-key attacks.
+//!
+//! NOTE: currently uses standard BLS aggregation without coefficient weighting, hence returns incorrect values compared to go-f3.
+//!
+use crate::verifier::BLSError;
+use bls_signatures::{PublicKey, Serialize, Signature};
+use bls12_381::{G1Projective, G2Projective};
+use filecoin_f3_gpbft::PubKey;
+
+/// BDN aggregation context for managing signature and public key aggregation
+pub struct BDNAggregation {
+    pub_keys: Vec<PubKey>,
+}
+
+impl BDNAggregation {
+    pub fn new(pub_keys: &[PubKey]) -> Result<Self, BLSError> {
+        if pub_keys.is_empty() {
+            return Err(BLSError::EmptyPublicKeys);
+        }
+
+        Ok(Self {
+            pub_keys: pub_keys.to_vec(),
+        })
+    }
+
+    /// Aggregates signatures using standard BLS aggregation
+    /// TODO: Implement proper BDN coefficient weighting
```

Comment:
Let's have a tracking issue for every TODO in the code. It's improbable anyone will pick it up by casually skimming through the code.

Along the lines of `TODO: Implement proper BDN coefficient weighting https://github.com/ChainSafe/rust-f3/issues/1234.

Reply:
Fixed here: https://github.com/ChainSafe/rust-f3/pull/27/commits/70bf5a53da44097f73b80f07ed7ac5107263c446